rfcs/20200106-prelu/README.md

```diff
@@ -0,0 +1,76 @@
+# Introducing PReLu support in OneDnn
+
+## Introduction
+
+This paper https://arxiv.org/abs/1502.01852 introduced PReLU activation function.
+Request comes from Caffe users: "PRelu is used in some POR topologies, such as
+VNET, SRGAN. It takes most of time if the original implementation, is not good
+optimized. For example VNET's train perf with latest Intel Caffe is 2.4s, PReLU
+(has optimized by openmp) spends 429ms (not including reorder time), 17% of total
+time. If we uses MKLDNN ReLU to make projection of the perf, it will come to
+~1.8s." Some other data scientist in discussion mentioned that he encouter PRelu
+issue when he tried to optimize the inference pipeline of a popular face-related
+application - PRelu tooks 47 % of time in his application.
+
+Currently in oneDnn we support ReLu activation function in eltwise operation:
+https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html. LeakyReLu can be
+achieved by specifying non zero alpha parameter. PReLu is specific variant of
+LeakyRelu, where alpha is learnable parameter. It cames with two flavours:
+- channel-wise PReLU - where number of trained alpha parameters is equal to
+channel number
+- channel-shared PReLu - where there is only one trainable alpha
+
+All popular framworks supports PReLu:
+- Caffe - https://caffe.berkeleyvision.org/tutorial/layers/prelu.html
+- Keras/tensorflow - https://www.tensorflow.org/api_docs/python/tf/keras/layers/PReLU
+- PyTorch - https://pytorch.org/docs/master/generated/torch.nn.PReLU.html
+- MxNet - https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.nn.PReLU.html
+
+Example of calculations forward/backward can be found: https://bit.ly/2TXgvzT
+
+## Proposal
+
+1. Option A - create new dedicated primitive PReLu:
+
+``` cpp
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_prelu.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training,
+    /// #dnnl_forward_inference, #dnnl_backward, and #dnnl_backward_data.
+    dnnl_prop_kind_t prop_kind;
+    /// The kind of eltwise algorithm.
+    /// #dnnl_prelu_channel_wise,
+    /// #dnnl_prelu_channel_shared.
+    dnnl_alg_kind_t alg_kind;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t data_desc;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t alpha_desc;
+    /// Source and destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_data_desc;
+    /// Alpha gradient memory descriptor.
+    dnnl_memory_desc_t diff_alpha_desc;
+
+} dnnl_prelu_desc_t;
```

Comment:
Is PReLU the only parameterized eltwise operation of interest? Does it make sense to have a more generic parameterized eltwise primitive (or at least an API that allows for extension to new parameterized eltwise operations, so that a new primitive does not need to be added for the next one)?

Reply:
Currently I have no idea. I will check frameworks / contact people and go back to you with answer.

Reply:
I received feedback from PyTorch team: "For PyTorch side, I made some checks , but not find such eltwise ops which have learnable parameters."
I checked Caffe and I don't see also such primitives https://github.com/BVLC/caffe/tree/master/src/caffe/layers

Reply:
Answer from tensorflow team: "I checked activation functions in TF, PRelu is the only one that has learnable parameter"